sql_grader/xblocks.py

```diff
@@ -0,0 +1,119 @@
+"""
+A custom submission type to evaluate SQL queries
+"""
+import logging
+import os
+import pkg_resources
+
+from xblock.core import XBlock
+from xblockutils.resources import ResourceLoader
+from xblockutils.studio_editable import StudioEditableXBlockMixin
+
+from .models import SqlProblem
+from .mixins.dates import EnforceDueDates
+from .mixins.fragment import XBlockFragmentBuilderMixin
+from .mixins.grading import Scorable
+from .mixins.scenario import XBlockWorkbenchMixin
+from .models import XBlockDataMixin
+
+LOGGER = logging.getLogger(__name__)
+
+
+def _(text):
+    """
+    Mock translation helper
+    """
+    return text
+
+
+def resource_string(path):
+    """
+    Handy helper for getting resources from our kit
+    """
+    data = pkg_resources.resource_string(__name__, path)
+    return data.decode('utf8')
+
+
+def all_datasets():
+    dataset_directory = pkg_resources.resource_filename('sql_grader', 'datasets')
+    for dirname, _, files in os.walk(dataset_directory):
+        for fname in files:
+            if fname.endswith('.sql'):
+                fname = fname[:-4]
+                yield fname
+
+
+def create_database(database_name):
+    pathname = 'datasets/' + database_name + '.sql'
+    contents = resource_string(pathname)
+    # print(contents)
+    database = SqlProblem.create_database_from_sql(contents)
+    return database
+
+
+DATABASES = {
```

Comment:
Load a copy of all databases into memory at startup time, to avoid overhead and disk access on each attempt/load.

We'll clone these on a per-request basis.

Reply:
What is this overhead, btw? FWIW, if it's small enough where it wouldn't be noticed, I'd be inclined to keep it simpler, recompute each time, and not have thread-level persistent state. If it's significantly larger, I'd prefer to keep it out of the module level init so that it doesn't factor into startup time for Studio/LMS (which is already seconds long–we end up doing an audit every year or so to cut stuff out). If we're talking a few tens of ms, please just do it in the simplest, least-dangerous way possible (i.e. do all the copying in the handler with no caching).

Reply:
The overhead is disk I/O.

I haven't run explicit benchmarks, but could.

Instinctively, I wanted to avoid the disk access on each page load/submit, as a best practice.

I've factored out the two user-scoped fields from the class (they were design relics), so I'm not too concerned about state persistence now; these objects are effectively read-only and used only to clone each per-load/submit.

That said, I can still refactor, if you'd prefer ;)

Reply:
Yeah, let's not worry about it for now.

Reply:
As in, let's keep it as-is and come back to it if it turns out to be a problem.